src/algebra/symmeterised.lean

```diff
@@ -0,0 +1,227 @@
+
+instance [nontrivial α] : nontrivial αˢʸᵐ := sym_injective.nontrivial
+instance [inhabited α] : inhabited αˢʸᵐ := ⟨sym default⟩
+instance [subsingleton α] : subsingleton αˢʸᵐ := unsym_injective.subsingleton
+instance [unique α] : unique αˢʸᵐ := unique.mk' _
+instance [is_empty α] : is_empty αˢʸᵐ := function.is_empty unsym
+
+
+instance [has_zero α] : has_zero (αˢʸᵐ) := { zero := sym 0 }
+
+instance [has_one α] : has_one αˢʸᵐ := { one := sym 1 }
+
+instance [has_add α] : has_add αˢʸᵐ :=
+{ add := λ a b, sym (unsym a + unsym b) }
+
+instance [has_sub α] : has_sub αˢʸᵐ := { sub := λ a b, sym (unsym a - unsym b) }
+
+instance [has_neg α] : has_neg αˢʸᵐ :=
+{ neg := λ a, sym (-unsym a) }
+
+/- Introduce the symmetrised multiplication-/
+instance [has_add α] [has_mul α] [has_one α] [invertible (2 : α)] : has_mul(αˢʸᵐ) :=
+{ mul := λ a b, sym (⅟2 * (unsym a * unsym b + unsym b * unsym a)) }
+
+@[to_additive] instance [has_inv α] : has_inv αˢʸᵐ :=
+{ inv := λ a, sym $ (unsym a)⁻¹ }
+
+instance (R : Type*) [has_scalar R α] : has_scalar R αˢʸᵐ :=
+{ smul := λ r a, sym (r • unsym a) }
+
+@[simp] lemma sym_zero [has_zero α] : sym (0 : α) = 0 := rfl
+@[simp] lemma unsym_zero [has_zero α] : unsym (0 : αˢʸᵐ) = 0 := rfl
+
+@[simp, to_additive] lemma sym_one [has_one α] : sym (1 : α) = 1 := rfl
+@[simp, to_additive] lemma unsym_one [has_one α] : unsym (1 : αˢʸᵐ) = 1 := rfl
+
+@[simp] lemma sym_add [has_add α] (a b : α) : sym (a + b) = sym a + sym b := rfl
+@[simp] lemma unsym_add [has_add α] (a b : αˢʸᵐ) : unsym (a + b) = unsym a + unsym b := rfl
+
+@[simp] lemma sym_neg [has_neg α] (a : α) : sym (-a) = -sym a := rfl
+@[simp] lemma unsym_neg [has_neg α] (a : αˢʸᵐ) : unsym (-a) = -unsym a := rfl
+
+@[simp, to_additive] lemma sym_inv [has_inv α] (a : α) : sym (a⁻¹) = (sym a)⁻¹ := rfl
+@[simp, to_additive] lemma unsym_inv [has_inv α] (a : αˢʸᵐ) : unsym (a⁻¹) = (unsym a)⁻¹ := rfl
+
+@[simp] lemma sym_sub [has_sub α] (a b : α) : sym (a - b) = sym a - sym b := rfl
+@[simp] lemma unsym_sub [has_sub α] (a b : αˢʸᵐ) : unsym (a - b) = unsym a - unsym b := rfl
+
+@[simp] lemma sym_smul {R : Type*} [has_scalar R α] (c : R) (a : α) : sym (c • a) = c • sym a := rfl
+@[simp] lemma unsym_smul {R : Type*} [has_scalar R α] (c : R) (a : αˢʸᵐ) :
+  unsym (c • a) = c • unsym a := rfl
+
+
+@[simp] lemma unsym_eq_zero_iff {α} [has_zero α] (a : αˢʸᵐ) : a.unsym = (0 : α) ↔ a = (0 : αˢʸᵐ) :=
+unsym_injective.eq_iff' rfl
+
+@[simp] lemma sym_eq_zero_iff [has_zero α] (a : α) : sym a = (0 : αˢʸᵐ) ↔ a = (0 : α) :=
+sym_injective.eq_iff' rfl
+
+lemma unsym_ne_zero_iff [has_zero α] (a : αˢʸᵐ) : a.unsym ≠ (0 : α) ↔ a ≠ (0 : αˢʸᵐ) :=
+not_congr $ unsym_eq_zero_iff a
+
+lemma sym_ne_zero_iff [has_zero α] (a : α) : sym a ≠ (0 : αˢʸᵐ) ↔ a ≠ (0 : α) :=
+not_congr $ sym_eq_zero_iff a
+
+@[simp] lemma unsym_eq_one_iff [has_one α] (a : αˢʸᵐ) : a.unsym = 1 ↔ a = 1 :=
+unsym_injective.eq_iff' rfl
+
+@[simp] lemma sym_eq_one_iff [has_one α] (a : α) : sym a = 1 ↔ a = 1 :=
```

Comment:
`to_additive` will generate all the zero lemmas from the one lemmas for you

Reply:
Done, thanks!